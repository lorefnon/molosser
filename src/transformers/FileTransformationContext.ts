import * as t from "@babel/types";
import * as path from "path";
import { some, upperFirst, capitalize } from "lodash";
import { Maybe } from "../util-types";
import { CompilationError } from "../CompilationError";

export class FileTransformationContext<ErrT = CompilationError> {
    public topLevelStatements: t.Statement[] = [
        t.importDeclaration([t.importNamespaceSpecifier(t.identifier("React"))], t.stringLiteral("react")),
    ];
    public errors: ErrT[] = [];

    constructor(public filePath?: string, public _defaultExportName?: string) {}

    shouldGenerate() {
        return !some(this.errors, { isFatal: true });
    }

    get defaultExportName(): Maybe<string> {
        if (this._defaultExportName) return this._defaultExportName;
        if (this.filePath) {
            const fileName = path.basename(this.filePath);
            return upperFirst(capitalize(fileName.split(".")[0]!));
        }
    }
}
