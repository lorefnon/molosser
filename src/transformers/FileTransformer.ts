import generateJS from '@babel/generator';
import * as t from "@babel/types";
import prettier from 'prettier';
// @ts-ignore
import pugLexer from 'pug-lexer';
// @ts-ignore
import pugParser from 'pug-parser';

import * as Pug from '../pug';
import { PugASTTransformer } from './PugASTTransformer';
import { Transformer } from './Transformer';
import { ErrorCode } from '../CompilationError';

export const parsePug = (input: string, context: {filePath?: string} = {}) => {
    const tokens = pugLexer(input, { filename: context.filePath });
    return pugParser(tokens, { filename: context.filePath, src: input });
}

export class FileTransfomer extends Transformer<string, string> {
    transform() {
        const ast = this.parseAST();
        if (!ast) return;
        const transformer = new PugASTTransformer(ast, this.context);
        transformer.transform();
        const statements = transformer.output;
        if (!statements) return;
        this.output = this.serialize(statements);
    }
    parseAST() {
        let ast;
        try {
            ast = parsePug(this.input, this.context);
        } catch (e) {
            this.pushError({
                code: ErrorCode.PugParseError,
                originalError: e,
                reasons: ["Failed to parse Pug Template"],
                isFatal: true,
            });
            return;
        }
        if (!Pug.isType<Pug.BlockNode>(ast, Pug.Type.Block)) {
            this.pushError({
                code: ErrorCode.UnsupportedSyntaxError,
                reasons: ["Expected top level to be a block"],
                column: 1,
                line: 1,
                isFatal: true,
            });
            return;
        }
        return ast;
    }
    serialize(statements: t.Statement[]) {
        let result: string;
        try {
            result = statements.map(n => generateJS(n).code).join("\n\n");
        } catch (e) {
            this.pushError({
                code: ErrorCode.GenerationError,
                reasons: ["Failed to generate javascript from transformed AST"],
                maybeBug: true,
                originalError: e,
                isFatal: true,
            });
            return;
        }
        try {
            result = prettier.format(result);
        } catch (e) {
            this.pushError({
                code: ErrorCode.GenerationError,
                reasons: ["Failed to prettify generated javascript"],
                isFatal: false,
            });
        }
        return result;
    }
}
