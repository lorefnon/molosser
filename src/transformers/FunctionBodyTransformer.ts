import * as t from "@babel/types";

import { Transformer } from "./Transformer";
import { first } from "lodash";
import { ErrorCode } from "../CompilationError";

export class FunctionBodyTransformer extends Transformer<t.Node[], t.Statement[]> {
    transform(): void {
        let precedingStatements: t.Statement[] = [];
        let jsxElements: (t.JSXElement | t.JSXExpressionContainer | t.JSXText)[] = [];
        for (let i = 0; i < this.input.length; i++) {
            const node = this.input[i];
            if (t.isStatement(node)) {
                if (jsxElements.length === 0) {
                    precedingStatements.push(node);
                } else {
                    const bodyStatements: t.Statement[] = [];
                    const transformer = new FunctionBodyTransformer(this.input.slice(i), this.context);
                    transformer.transform();
                    if (transformer.output) bodyStatements.push(...transformer.output);
                    jsxElements.push(
                        t.jsxExpressionContainer(
                            t.callExpression(t.arrowFunctionExpression([], t.blockStatement(bodyStatements)), []),
                        ),
                    );
                    break;
                }
            } else if (t.isJSXElement(node) || t.isJSXText(node)) {
                jsxElements.push(node);
            } else if (t.isExpression(node)) {
                jsxElements.push(t.jsxExpressionContainer(node));
            } else {
                this.pushError({
                    code: ErrorCode.UnsupportedSyntaxError,
                    isFatal: false,
                    reasons: [`Encountered unexpected node of type ${node.type}`],
                });
            }
        }
        if (jsxElements.length === 0) this.output = precedingStatements;
        if (jsxElements.length === 1) {
            const soloEl = first(jsxElements)!;
            if (t.isJSXElement(soloEl)) this.output = [...precedingStatements, t.returnStatement(soloEl)];
        }
        this.output = this.output || [
            ...precedingStatements,
            t.returnStatement(t.jsxFragment(t.jsxOpeningFragment(), t.jsxClosingFragment(), jsxElements)),
        ];
    }
}
