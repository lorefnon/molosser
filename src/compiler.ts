import { FileTransformationContext } from './transformers/FileTransformationContext';
import { FileTransfomer } from './transformers/FileTransformer';

export interface CompilationOptions {
    filePath?: string;
    defaultExportName?: string;
}

export const transpile = (
    inputContent: string,
    options: CompilationOptions = {}
) => {
    const context = new FileTransformationContext(options.filePath, options.defaultExportName);
    const transformer = new FileTransfomer(inputContent, context);
    transformer.transform();
    return {
        result: transformer.output,
        errors: context.errors
    };
};