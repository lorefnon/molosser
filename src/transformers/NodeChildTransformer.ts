import * as Pug from "../pug";
import * as t from "@babel/types";
import { Transformer } from "./Transformer";
import { TagBlockNodeTransformer } from "./TagBlockNodeTransformer";
import { NodeChildrenTransformer } from "./NodeChildrenTransformer";

export type JSXChild = (t.JSXText | t.JSXExpressionContainer | t.JSXSpreadChild | t.JSXElement | t.JSXFragment);
export type JSXNode = (t.JSXElement | t.JSXFragment);

export class NodeChildTransformer extends Transformer<
    Pug.Node,
    JSXNode
> {
    transform(): void {
        let nodes: JSXChild[] = [];
        if (Pug.isType<Pug.BlockNode>(this.input, Pug.Type.Block)) {
            nodes = this.delegateTo(TagBlockNodeTransformer, this.input) || nodes;
        } else {
            nodes = this.delegateTo(NodeChildrenTransformer, [this.input]) || nodes;
        }
        if (nodes.length === 1) {
            const firstNode = nodes[0];
            if (t.isJSXFragment(firstNode) || t.isJSXElement(firstNode)) {
                this.output = firstNode;
            }
        }
        this.output = this.output || t.jsxFragment(t.jsxOpeningFragment(), t.jsxClosingFragment(), nodes);
    }
}
