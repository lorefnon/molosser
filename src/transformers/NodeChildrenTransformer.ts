import * as Pug from "../pug";
import * as t from "@babel/types";
import { Transformer } from "./Transformer";
import { PugNodeTransformer } from "./PugNodeTransformer";
import { FunctionBodyTransformer } from "./FunctionBodyTransformer";
import { flatten, compact } from "lodash";
import { JSXChild } from "./NodeChildTransformer";

// Transforms to children which can be embedded within JSX
export class NodeChildrenTransformer extends Transformer<Pug.Node[], JSXChild[]> {
    transform(): void {
        this.output = [];
        const transformedNodes = flatten(compact(this.input.map(n => this.delegateTo(PugNodeTransformer, n))));
        for (let i = 0; i < transformedNodes.length; i++) {
            const childNode = transformedNodes[i];
            if (t.isJSXElement(childNode) || t.isJSXText(childNode)) this.output.push(childNode);
            else if (t.isExpression(childNode) || t.isIdentifier(childNode))
                this.output.push(t.jsxExpressionContainer(childNode));
            else {
                this.output.push(
                    t.jsxExpressionContainer(
                        t.callExpression(
                            t.arrowFunctionExpression(
                                [],
                                t.blockStatement(
                                    this.delegateTo(FunctionBodyTransformer, transformedNodes.slice(i)) || [],
                                ),
                            ),
                            [],
                        ),
                    ),
                );
                break;
            }
        }
    }
}
