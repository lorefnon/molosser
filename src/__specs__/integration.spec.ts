import { transpile} from "../compiler";
import dedent from "dedent";
import { isEmpty } from "lodash";
import { parsePug } from "../transformers/FileTransformer";

test("Tag composition", () => {
    const template = dedent`
    #bar.foo
        .foo__node.foo__node--aspect(data-bar="baz")
            | Hello world`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(transpiledR).toMatchSnapshot();
    expect(isEmpty(transpiledR.errors)).toBe(true);
});


test("Interpolated Code", () => {
    const template = dedent`
    - const foo = "bar";
    #bar.foo
        .foo__node.foo__node--aspect(data-bar="baz")= foo
    div.hello
        = \`Hello \${foo}\``;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Interleaved unbuffered code", () => {
    const template = dedent`
    - const foo = "bar";
    #bar.foo
        .foo__node.foo__node--aspect Test
    - const bar = "baz";
    div.hello
        = \`Hello \${foo} \${bar}\``;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Top level scripts", () => {
    const template = dedent`
    script.
        import SomeTag from "./SomeTag";
        
    div.hello
        SomeTag(name=props.foo)`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template);
    console.log(transpiledR.errors);
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Props interface detection", () => {
    const template = dedent`
    script.
        interface ITestComponentProps {
            foo: string;
        }
        
    div.hello
        SomeTag(name=props.foo)`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR: any = transpile(template, {
        defaultExportName: "TestComponent",
    });
    console.log(transpiledR.errors);
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

