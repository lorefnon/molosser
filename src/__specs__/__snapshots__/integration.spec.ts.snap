// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assigning react elements 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "block": Object {
        "line": 2,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 6,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'hello'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "column": 7,
                  "line": 3,
                  "type": "Text",
                  "val": "World",
                },
              ],
              "type": "Block",
            },
            "column": 3,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const foo =",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 5,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 5,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 13,
                "line": 6,
                "mustEscape": true,
                "name": "name",
                "val": "foo",
              },
            ],
            "block": Object {
              "line": 6,
              "nodes": Array [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 6,
            "name": "SomeTag",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 5,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Assigning react elements 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

export default function TestComponent(props: any) {
  const foo = <div className={\\"hello\\"}>World</div>;
  return (
    <div className={\\"hello\\"}>
      <SomeTag name={foo} />
    </div>
  );
}
",
}
`;

exports[`Interleaved unbuffered code 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const foo = \\"bar\\";",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 1,
          "line": 2,
          "mustEscape": false,
          "name": "id",
          "val": "'bar'",
        },
        Object {
          "column": 5,
          "line": 2,
          "mustEscape": false,
          "name": "class",
          "val": "'foo'",
        },
      ],
      "block": Object {
        "line": 2,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 3,
                "mustEscape": false,
                "name": "class",
                "val": "'foo__node'",
              },
              Object {
                "column": 15,
                "line": 3,
                "mustEscape": false,
                "name": "class",
                "val": "'foo__node--aspect'",
              },
            ],
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "column": 34,
                  "line": 3,
                  "type": "Text",
                  "val": "Test",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 3,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 2,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
    Object {
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 4,
      "mustEscape": false,
      "type": "Code",
      "val": "const bar = \\"baz\\";",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 5,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 5,
        "nodes": Array [
          Object {
            "buffer": true,
            "column": 5,
            "isInline": false,
            "line": 6,
            "mustEscape": true,
            "type": "Code",
            "val": "\`Hello \\\\\${foo} \\\\\${bar}\`",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 5,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Interleaved unbuffered code 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

export default function SampleComponent(props: any) {
  const foo = \\"bar\\";
  return (
    <>
      <div id={\\"bar\\"} className={\\"foo\\"}>
        <div className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}>Test</div>
      </div>
      {(() => {
        const bar = \\"baz\\";
        return <div className={\\"hello\\"}>{\`Hello \\\\\${foo} \\\\\${bar}\`}</div>;
      })()}
    </>
  );
}
",
}
`;

exports[`Internal functions returning JSX 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "block": Object {
        "line": 2,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 6,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'hello'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "column": 7,
                  "line": 3,
                  "type": "Text",
                  "val": "World",
                },
              ],
              "type": "Block",
            },
            "column": 3,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const SomeTag = () =>",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 5,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 5,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 13,
                "line": 6,
                "mustEscape": true,
                "name": "name",
                "val": "foo",
              },
            ],
            "block": Object {
              "line": 6,
              "nodes": Array [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 6,
            "name": "SomeTag",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 5,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Internal functions returning JSX 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

export default function TestComponent(props: any) {
  const SomeTag = () => <div className={\\"hello\\"}>World</div>;

  return (
    <div className={\\"hello\\"}>
      <SomeTag name={foo} />
    </div>
  );
}
",
}
`;

exports[`Interpolated Code 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const foo = \\"bar\\";",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 1,
          "line": 2,
          "mustEscape": false,
          "name": "id",
          "val": "'bar'",
        },
        Object {
          "column": 5,
          "line": 2,
          "mustEscape": false,
          "name": "class",
          "val": "'foo'",
        },
      ],
      "block": Object {
        "line": 2,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 3,
                "mustEscape": false,
                "name": "class",
                "val": "'foo__node'",
              },
              Object {
                "column": 15,
                "line": 3,
                "mustEscape": false,
                "name": "class",
                "val": "'foo__node--aspect'",
              },
              Object {
                "column": 34,
                "line": 3,
                "mustEscape": true,
                "name": "data-bar",
                "val": "\\"baz\\"",
              },
            ],
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "buffer": true,
                  "column": 49,
                  "isInline": true,
                  "line": 3,
                  "mustEscape": true,
                  "type": "Code",
                  "val": "foo",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 3,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 2,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 4,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 4,
        "nodes": Array [
          Object {
            "buffer": true,
            "column": 5,
            "isInline": false,
            "line": 5,
            "mustEscape": true,
            "type": "Code",
            "val": "\`Hello \\\\\${foo}\`",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 4,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Interpolated Code 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

export default function SampleComponent(props: any) {
  const foo = \\"bar\\";
  return (
    <>
      <div id={\\"bar\\"} className={\\"foo\\"}>
        <div
          data-bar={\\"baz\\"}
          className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}
        >
          {foo}
        </div>
      </div>
      <div className={\\"hello\\"}>{\`Hello \\\\\${foo}\`}</div>
    </>
  );
}
",
}
`;

exports[`Lifted function components 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 8,
          "line": 1,
          "mustEscape": true,
          "name": "type",
          "val": "\\"text/molosser\\"",
        },
      ],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "buffer": false,
                  "column": 9,
                  "isInline": false,
                  "line": 3,
                  "mustEscape": false,
                  "type": "Code",
                  "val": "const [currentCount, setCount] = useState(0)",
                },
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 12,
                      "line": 4,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'hello'",
                    },
                  ],
                  "block": Object {
                    "line": 4,
                    "nodes": Array [
                      Object {
                        "buffer": true,
                        "column": 13,
                        "isInline": false,
                        "line": 5,
                        "mustEscape": true,
                        "type": "Code",
                        "val": "currentCount",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "isInline": false,
                  "line": 4,
                  "name": "div",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": false,
            "type": "Code",
            "val": "const SomeTag = () =>",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "script",
      "selfClosing": false,
      "type": "Tag",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 7,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 7,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 13,
                "line": 8,
                "mustEscape": true,
                "name": "name",
                "val": "foo",
              },
            ],
            "block": Object {
              "line": 8,
              "nodes": Array [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 8,
            "name": "SomeTag",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 7,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Lifted function components 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

const SomeTag = () => {
  const [currentCount, setCount] = useState(0);
  return <div className={\\"hello\\"}>{currentCount}</div>;
};

export default function TestComponent(props: any) {
  return (
    <div className={\\"hello\\"}>
      <SomeTag name={foo} />
    </div>
  );
}
",
}
`;

exports[`Props interface detection 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 5,
            "line": 2,
            "type": "Text",
            "val": "interface ITestComponentProps {",
          },
          Object {
            "column": 1,
            "line": 3,
            "type": "Text",
            "val": "
",
          },
          Object {
            "column": 5,
            "line": 3,
            "type": "Text",
            "val": "    foo: string;",
          },
          Object {
            "column": 1,
            "line": 4,
            "type": "Text",
            "val": "
",
          },
          Object {
            "column": 5,
            "line": 4,
            "type": "Text",
            "val": "}",
          },
          Object {
            "column": 1,
            "line": 5,
            "type": "Text",
            "val": "
",
          },
          Object {
            "column": 5,
            "line": 5,
            "type": "Text",
            "val": "",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "script",
      "selfClosing": false,
      "textOnly": true,
      "type": "Tag",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 6,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 6,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 13,
                "line": 7,
                "mustEscape": true,
                "name": "name",
                "val": "props.foo",
              },
            ],
            "block": Object {
              "line": 7,
              "nodes": Array [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 7,
            "name": "SomeTag",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 6,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Props interface detection 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

interface ITestComponentProps {
  foo: string;
}

export default function TestComponent(props: ITestComponentProps) {
  return (
    <div className={\\"hello\\"}>
      <SomeTag name={props.foo} />
    </div>
  );
}
",
}
`;

exports[`Render props 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 11,
          "line": 1,
          "mustEscape": false,
          "name": "class",
          "val": "'Consumer'",
        },
      ],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "alternate": Object {
                    "line": 6,
                    "nodes": Array [
                      Object {
                        "attributeBlocks": Array [],
                        "attrs": Array [
                          Object {
                            "column": 13,
                            "line": 6,
                            "mustEscape": false,
                            "name": "id",
                            "val": "'workspace'",
                          },
                        ],
                        "block": Object {
                          "line": 6,
                          "nodes": Array [
                            Object {
                              "column": 24,
                              "line": 6,
                              "type": "Text",
                              "val": "Failure",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 6,
                        "name": "div",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "consequent": Object {
                    "line": 4,
                    "nodes": Array [
                      Object {
                        "attributeBlocks": Array [],
                        "attrs": Array [
                          Object {
                            "column": 13,
                            "line": 4,
                            "mustEscape": false,
                            "name": "id",
                            "val": "'workspace'",
                          },
                        ],
                        "block": Object {
                          "line": 4,
                          "nodes": Array [
                            Object {
                              "column": 24,
                              "line": 4,
                              "type": "Text",
                              "val": "Success",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 4,
                        "name": "div",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "line": 3,
                  "test": "app.repositoryRoot",
                  "type": "Conditional",
                },
              ],
              "type": "Block",
            },
            "buffer": true,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": true,
            "type": "Code",
            "val": "(app) =>",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "AppContext",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Render props 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

export default function(props: any) {
  return (
    <AppContext className={\\"Consumer\\"}>
      {app =>
        app.repositoryRoot ? (
          <div id={\\"workspace\\"}>Success</div>
        ) : (
          <div id={\\"workspace\\"}>Failure</div>
        )
      }
    </AppContext>
  );
}
",
}
`;

exports[`Tag composition 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 1,
          "line": 1,
          "mustEscape": false,
          "name": "id",
          "val": "'bar'",
        },
        Object {
          "column": 5,
          "line": 1,
          "mustEscape": false,
          "name": "class",
          "val": "'foo'",
        },
      ],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'foo__node'",
              },
              Object {
                "column": 15,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'foo__node--aspect'",
              },
              Object {
                "column": 34,
                "line": 2,
                "mustEscape": true,
                "name": "data-bar",
                "val": "\\"baz\\"",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "column": 11,
                  "line": 3,
                  "type": "Text",
                  "val": "Hello world",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Tag composition 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

export default function SampleComponent(props: any) {
  return (
    <div id={\\"bar\\"} className={\\"foo\\"}>
      <div
        data-bar={\\"baz\\"}
        className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}
      >
        Hello world
      </div>
    </div>
  );
}
",
}
`;

exports[`Top level scripts 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 5,
            "line": 2,
            "type": "Text",
            "val": "import SomeTag from \\"./SomeTag\\";",
          },
          Object {
            "column": 1,
            "line": 3,
            "type": "Text",
            "val": "
",
          },
          Object {
            "column": 5,
            "line": 3,
            "type": "Text",
            "val": "",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "script",
      "selfClosing": false,
      "textOnly": true,
      "type": "Tag",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [
        Object {
          "column": 4,
          "line": 4,
          "mustEscape": false,
          "name": "class",
          "val": "'hello'",
        },
      ],
      "block": Object {
        "line": 4,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 13,
                "line": 5,
                "mustEscape": true,
                "name": "name",
                "val": "props.foo",
              },
            ],
            "block": Object {
              "line": 5,
              "nodes": Array [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 5,
            "name": "SomeTag",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 4,
      "name": "div",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Top level scripts 2`] = `
Object {
  "errors": Array [],
  "result": "import * as React from \\"react\\";

import SomeTag from \\"./SomeTag\\";

export default function(props: any) {
  return (
    <div className={\\"hello\\"}>
      <SomeTag name={props.foo} />
    </div>
  );
}
",
}
`;
